pkg/action/executor/move_util.go

```diff
@@ -36,63 +40,141 @@ func copyPodInfo(oldPod, newPod *api.Pod) {
-		id, pod.Spec.NodeName, nodeName)
+	// 0. get the latest Pod
+	var err error
+	var pod *api.Pod
+	if pod, err = podClient.Get(tpod.Name, metav1.GetOptions{}); err != nil {
+		glog.Errorf("Failed to get latest pod %v/%v: %v", tpod.Namespace, tpod.Name, err)
+		return nil, err
+	}
+	//NOTE: do deep-copy if the original pod may be modified outside this function
+	labels := pod.Labels
 
-	npod := &api.Pod{}
-	copyPodInfo(pod, npod)
-	npod.Spec.NodeName = nodeName
+	//1. create a clone pod--podC of the original pod--podA
+	npod, err := createClonePod(client, pod, nodeName)
+	if err != nil {
+		glog.Errorf("Move pod failed: failed to create a clone pod: %v", err)
+		return nil, err
+	}
+
+	//delete the clone pod if this action fails
+	flag := false
+	defer func() {
+		if !flag {
+			glog.Errorf("Move pod failed, begin to delete cloned pod: %v/%v", npod.Namespace, npod.Name)
+			delOpt := &metav1.DeleteOptions{}
+			podClient.Delete(npod.Name, delOpt)
+		}
+	}()
 
-	//2. kill original pod
-	grace := calcGracePeriod(pod)
-	delOption := &metav1.DeleteOptions{GracePeriodSeconds: &grace}
-	err := podClient.Delete(pod.Name, delOption)
+	//1.2 wait until podC gets ready
+	err = waitForReady(client, npod.Namespace, npod.Name, nodeName, retryNum)
 	if err != nil {
-		err = fmt.Errorf("move-failed: failed to delete original pod-%v: %v",
-			id, err)
-		glog.Error(err)
+		glog.Errorf("Wait for cloned Pod ready timeout: %v", err)
 		return nil, err
 	}
 
-	//3. create (and bind) the new Pod
-	time.Sleep(time.Duration(grace)*time.Second + defaultMoreGrace) //wait for the previous pod to be cleaned up.
-	interval := defaultPodCreateSleep
-	timeout := time.Duration(retryNum+1) * interval
-	err = util.RetryDuring(retryNum, timeout, interval, func() error {
-		_, inerr := podClient.Create(npod)
-		return inerr
-	})
+	//2. delete the original pod--podA
+	delOpt := &metav1.DeleteOptions{}
+	if err := podClient.Delete(pod.Name, delOpt); err != nil {
+		glog.Warningf("Move pod warning: failed to delete original pod: %v", err)
+	}
+
+	//3. add labels to podC
+	xpod, err := podClient.Get(npod.Name, metav1.GetOptions{})
 	if err != nil {
-		err = fmt.Errorf("move-failed: failed to create new pod-%v: %v",
-			id, err)
-		glog.Error(err)
+		glog.Errorf("Move pod failed: failed to get the cloned pod: %v", err)
+		return nil, err
+	}
+
+	//TODO: compare resourceVersion of xpod and npod before updating
+	if (labels != nil) && len(labels) > 0 {
+		xpod.Labels = labels
+		if _, err := podClient.Update(xpod); err != nil {
+			glog.Errorf("Move pod failed: failed to update labels for cloned pod: %v", err)
+			return nil, err
+		}
+	}
+
+	flag = true
+	return xpod, nil
+}
+
+func createClonePod(client *kclient.Clientset, pod *api.Pod, nodeName string) (*api.Pod, error) {
+	npod := &api.Pod{}
+	copyPodWithoutLabel(pod, npod)
+	npod.Spec.NodeName = nodeName
+	npod.Name = genNewPodName(pod.Name)
+	util.AddAnnotation(npod, TurboActionAnnotationKey, TurboMoveAnnotationValue)
```

Comment:
Question: if we have a pod with both turbo move and resize actions applied, we will only see one annotation for the latest action, is it desired?

Reply:
yes,  action annotation item is overwritten, so only the latest action info is saved.